sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -573,4 +573,109 @@ object DateTimeUtils {
+    } else if (remainYear < 330) {
+      remainYear / 4 - 2
+    } else {
+      remainYear / 4 - 3
+    }
+    cycles * (365 * 400 + 397) + remainYear * 365 + numLeaps
+  }
+
+  /**
+   * Add date and year-month interval.
+   * Returns a date value, expressed in days since 1.1.1970.
+   */
+  def dateAddYearMonthInterval(days: Int, months: Int): Int = {
+    val currentMonth = (getYear(days) - 1970) * 12 + getMonth(days) - 1 + months
+    val currentMonthInYear = if (currentMonth < 0) {
+      ((currentMonth % 12) + 12) % 12
+    } else {
+      currentMonth % 12
+    }
+    val currentYear = if (currentMonth < 0) (currentMonth / 12) - 1 else currentMonth / 12
+    val febDays = if (isLeapYear(1970 + currentYear)) 29 else 28
+    val daysForMonths = Seq(31, febDays, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
+    getDaysFromMonths(currentMonth) + math.min(
+      getDayOfMonth(days), daysForMonths(currentMonthInYear)) - 1
+  }
+
+  /**
+   * Add date and full interval.
+   * Returns a timestamp value, expressed in microseconds since 1.1.1970 00:00:00.
+   */
+  def dateAddFullInterval(days: Int, months: Int, microseconds: Long): Long = {
+    daysToMillis(dateAddYearMonthInterval(days, months)) * 1000L + microseconds
+  }
+
+  /**
+   * Add timestamp and full interval.
+   * Returns a timestamp value, expressed in microseconds since 1.1.1970 00:00:00.
+   */
+  def timestampAddFullInterval(micros: Long, months: Int, microseconds: Long): Long = {
+    val days = millisToDays(micros / 1000L)
+    dateAddFullInterval(days, months, microseconds) + micros - daysToMillis(days) * 1000L
+  }
+
+  /**
+   * Returns the last dayInMonth in the month it belongs to. The date is expressed
+   * in days since 1.1.1970. the return value starts from 1.
+   */
+  def getLastDayInMonthOfMonth(date: Int): Int = {
+    val month = getMonth(date)
+
+    val febDay = if (isLeapYear(getYear(date))) 29 else 28
+    val days = Seq(31, febDay, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
+    days(month - 1)
+  }
+
+  /**
+   * Returns number of months between time1 and time2. time1 and time2 are expressed in
+   * microseconds since 1.1.1970
+   */
+  def monthsBetween(time1: Long, time2: Long): Double = {
+    val millis1 = time1.asInstanceOf[Long] / 1000L
+    val millis2 = time2.asInstanceOf[Long] / 1000L
+    val date1 = millisToDays(millis1)
+    val date2 = millisToDays(millis2)
+    val dayInMonth1 = getDayOfMonth(date1)
+    val dayInMonth2 = getDayOfMonth(date2)
+    val lastDayMonth1 = getLastDayInMonthOfMonth(date1)
+    val lastDayMonth2 = getLastDayInMonthOfMonth(date2)
+    val months1 = getYear(date1) * 12 + getMonth(date1) - 1
+    val months2 = getYear(date2) * 12 + getMonth(date2) - 1
+    val timeInDay1 = time1 - daysToMillis(date1) * 1000L
+    val timeInDay2 = time2 - daysToMillis(date2) * 1000L
+    if (dayInMonth1 == dayInMonth2 || (lastDayMonth1 == dayInMonth1 &&
+      lastDayMonth2 == dayInMonth2) || dayInMonth2 >= lastDayMonth1) {
```

Comment:
inconsistent with Hive here for `dayInMonth2 >= lastDayMonth1`?
>  If date1 and date2 are either the same days of the month or both last days of months, then the result is always an integer. Otherwise the UDF calculates the fractional portion of the result based on a 31-day month ...